File: TaskManager/TaskManager_210042112/TaskManager_112/TaskManager_112/ICommand Interface/CommandInvoker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TaskManager_112
{
    public class CommandInvoker
    {
        private Stack<ICommand> commandHistory = new Stack<ICommand>();
        private Stack<ICommand> undoneCommands = new Stack<ICommand>();
        public void ExecuteCommand(ICommand command)
        {
            command.Execute();
            commandHistory.Push(command);
            undoneCommands.Clear(); 
        }
        public void Undo()
        {
            if (commandHistory.Count > 0)
            {
                ICommand command = commandHistory.Pop();
                command.Undo();
                undoneCommands.Push(command);
            }
        }
        public void Redo()
        {
            if (undoneCommands.Count > 0)
            {
                ICommand command = undoneCommands.Pop();
                command.Execute();
                commandHistory.Push(command);
            }
        }
    }
}

File: TaskManager/210042141_TaskManager/src/CommandInvoker.java
import java.util.Stack;

public class CommandInvoker {
    private Stack<Command> commandHistory = new Stack<>();
    private Stack<Command> redoHistory = new Stack<>();

    public void executeCommand(Command command) {
        command.execute();
        commandHistory.push(command);
        redoHistory.clear();
    }

    public void undo() {
        if (!commandHistory.isEmpty()) {
            Command command = commandHistory.pop();
            command.undo();
            redoHistory.push(command);
        }
    }

    public void redo() {
        if (!redoHistory.isEmpty()) {
            Command command = redoHistory.pop();
            command.execute();
            commandHistory.push(command);
        }
    }
}
